### DISCUSSION TITLE
Batch of questions

### DISCUSSION THREAD

**Discussion by alchem1ster** [AllenDang/giu#450] [Q&A] [answered]
Hello, folks!
I have used this to build simple (almost) static GUIs, but now I needed something more responsive. So, I have a list of questions that I couldn't find the answers to, either in the examples or the original imgui-go. I've been programming complex interfaces in Qt for a long time, and maybe that's why I have no understanding of how it should work for Immediate paradigm. 

<hr>

### 1. Switching the menu

<details>
<summary>main.go</summary>

```golang
package main

import (
	"fmt"

	"github.com/AllenDang/giu"
)

func btnClickMeClicked() {
	fmt.Println("Action!")
}

func loop() {
	giu.SingleWindowWithMenuBar().Layout(
		giu.MenuBar().Layout(
			giu.Menu("File").Layout(
				giu.MenuItem("Switch").OnClick(btnClickMeClicked),
				giu.Menu("Save as ...").Layout(
					giu.MenuItem("Excel file"),
					giu.MenuItem("CSV file"),
				),
			),
		),
		giu.Label("One line label"),
	)
}

func main() {
	w := giu.NewMasterWindow("Overview", 800, 400, 0)
	w.Run(loop)
}
```
</details>

**Question:** How can I make the "Save as" item (with whole layout) disappear when the "Switch" button is clicked and other menu items were added instead (e.g., "Open"). But when you press "Switch" again, everything should go back to the original version.

### 2. Change button function
**Question:** How can I change the onClick function of specific button from function of another button? For example, we have button_1, which outputs "1" when pressed, and button_2, which outputs "2". How do we make a button_3 that, when clicked, will swap buttons 1 and 2?

### 3. Show/Hide elements
**Question:** Can you show an example of hiding/displaying an element inside the MasterWindow? For example, hiding/displaying a slider area. For example, for an image previewer, the preview area will only appear when you select an image, not a folder or a graphic file. This is similar to my first question, but we are not talking about menus here, but any elements in general.

### 4. Non-blocking new window
**Question:** How to create a new window with a custom form to fill in (for example, login and password) when you click on some button so that it does not block interaction with the main program.

<hr>

Maybe there will somehow be links where all this can be read, I would appreciate it.  
In fact, there will be many more questions and sorry cause I don't know where to ask them.

**Comment by gucio321** [comment]
> Question: How can I make the "Save as" item (with whole layout) disappear when the "Switch" button is clicked and other menu items were added instead (e.g., "Open"). But when you press "Switch" again, everything should go back to the original version.

take a look on this:
<details><summary>main.go</summary>

```golang
package main

import (
	"fmt"

	"github.com/AllenDang/giu"
)

var isActive bool = false

func btnClickMeClicked() {
	isActive = !isActive
	fmt.Println("Action!")
}

func loop() {
	giu.SingleWindowWithMenuBar().Layout(
		giu.MenuBar().Layout(
			giu.Menu("File").Layout(
				giu.MenuItem("Switch").OnClick(btnClickMeClicked),
				giu.Custom(func() {
					if isActive {
						giu.Menu("Save as ...").Layout(
							giu.MenuItem("Excel file"),
							giu.MenuItem("CSV file"),
						).Build()
					}
				}),
			),
		),
		giu.Label("One line label"),
	)
}

func main() {
	w := giu.NewMasterWindow("Overview", 800, 400, 0)
	w.Run(loop)
}
```

</details>

**Comment by gucio321** [comment, marked as answer]
> Question: How can I change the onClick function of specific button from function of another button? For example, we have button_1, which outputs "1" when pressed, and button_2, which outputs "2". How do we make a button_3 that, when clicked, will swap buttons 1 and 2?

just like in Question 1, you need to add separated variable and spawn your buttens depending on its value:
<details><summary>main.go</summary>

```golang
package main

import (
	"fmt"

	"github.com/AllenDang/giu"
)

var isActive bool = false

func loop() {
	giu.SingleWindowWithMenuBar().Layout(
		giu.Button("button 3").OnClick(func() {
			isActive = !isActive
		}),
		giu.Condition(
			isActive,
			giu.Layout{
				giu.Button("button1").OnClick(func() { fmt.Println("1") }),
				giu.Button("button2").OnClick(func() { fmt.Println("2") }),
			},
			giu.Layout{},
		),
	)
}

func main() {
	w := giu.NewMasterWindow("Overview", 800, 400, 0)
	w.Run(loop)
}
```

</details>

**Reply by alchem1ster** [reply]
I meant that when you press button_3 the functions of buttons 1 and 2 should switch places (second now prints "1" and first "2").
And why in first situation you used giu.Custom() but giu.Condition() now?

**Reply by gucio321** [reply]
also, in case of your third question, you need to do it like above.
you need to understand how IMGUI works.
The `loop` function is called in each frame and it builds all the widgets. So you can easly manage, whether the item should be built in a frame or not. 
to do so, you have 2 ways:
1. use giu's ConditionWIdget
2. if you don't want to use Condition(...) or your conditions are more complicated (e.g. u have 3 cases) you can always use CustomWIdget with your own implementation of giu/imgui-go widgets's calls

**Reply by gucio321** [reply]
oh, we posted at the same time :grinning:

**Reply by gucio321** [reply]
> I meant that when you press button_3 the functions of buttons 1 and 2 should switch places (second now prints "1" and first "2").

hmm, well so I'd do like:
<details><summary>main.go</summary>

```golang
package main

import (
	"fmt"

	"github.com/AllenDang/giu"
)

var isActive bool = false

func fn1() {
	fmt.Println("1")
}

func fn2() {
	fmt.Println("2")
}

func loop() {
	giu.SingleWindowWithMenuBar().Layout(
		giu.Button("button 3").OnClick(func() {
			isActive = !isActive
		}),

		giu.Button("button1").OnClick(func() {
			if isActive {
				fn1()
				return
			}

			fn2()
		}),
		giu.Button("button2").OnClick(func() {
			if isActive {
				fn2()
				return
			}

			fn1()
		}),
	)
}

func main() {
	w := giu.NewMasterWindow("Overview", 800, 400, 0)
	w.Run(loop)
}
```

</details>

but this code dosn't look good :-(

> And why in first situation you used giu.Custom() but giu.Condition() now?

yah, giu.Condition is giu-api-styled widget and giu.Custom just allows you to run your own functions

**Reply by alchem1ster** [reply]
@gucio321 
Thank you so much! I think I'm beginning to understand how it works. But I still have a question about the last point. How to create a child (second) window? Let's say I have a window with a fixed size of 800x600, which does not fit all the elements, but I can not increase the size of the window for this purpose. Solution: create a new child window with other elements, but shared memory. For example, a request for a login and password. How to implement this?

**Reply by gucio321** [reply]
@alchem1ster it depends on your exact use case.
You can for example create a popupmodal window OR just create another WindowWidget.
for first, you can take a look on [`widgets`](https://github.com/AllenDang/giu/blob/master/examples/widgets/widgets.go) example 
![image](https://user-images.githubusercontent.com/73652197/155890509-34b34e1a-3de9-48f1-8c32-d7669384a80e.png)
for multiwindows, see `examples/multiwindow`

**Reply by gucio321** [reply]
btw, there is a nice impl of simple message-popups called MsgboxWidget

**Reply by alchem1ster** [reply]
I found only https://github.com/AllenDang/giu/wiki#messagebox did you mean this?
Or could you show any example of how to use MsgboxWidget? Thanks a lot, mate :)

**Reply by gucio321** [reply]
yah @alchem1ster if you're about to use Msgbox, you can do in a simillar way as on wiki, however the wiki is a bit out of data
CC @AllenDang 

well, there is an example for msgbox:
<details><summary>msgbox.go</summary>

```golang
package main

import (
	"fmt"

	"github.com/AllenDang/giu"
)

func loop() {
	giu.Window("window").Layout(
		giu.PrepareMsgbox(),
		giu.Button("click me to see message box").OnClick(func() {
			giu.Msgbox("Info", "I'm a msgbox. press OK to close me")
		}),
		giu.Button("show yes-no dialog").OnClick(func() {
			giu.Msgbox("Question", "I'm yes-no dialog. Please take an action").
				Buttons(giu.MsgboxButtonsYesNo).
				ResultCallback(func(result giu.DialogResult) {
					switch result {
					case giu.DialogResultYes:
						fmt.Println("Yes clicked")
					case giu.DialogResultNo:
						fmt.Println("No clicked")
					}
				})
		}),
		giu.Button("show ok-cancel dialog").OnClick(func() {
			giu.Msgbox("ok-cancel", "I'm ok-cancel dialog").Buttons(giu.MsgboxButtonsOkCancel)
		}),
	)
}

func main() {
	wnd := giu.NewMasterWindow("Msg box demo", 640, 480, 0)
	wnd.Run(loop)
}
```

</details>

**Reply by alchem1ster** [reply]
Is it possible to add input fields or other elements there? For example, tables or sliders? Or another layout?

**Reply by gucio321** [reply]
sure! in the following demo you can put anything, you want instead of a label
```golang
package main

import "github.com/AllenDang/giu"

func loop() {
	giu.SingleWindow().Layout(
		giu.PopupModal("popup").Layout(
			giu.Label("any layout you want"),
			giu.Button("Close me").OnClick(func() {
				giu.CloseCurrentPopup()
			}),
		),
		giu.Button("Open popup").OnClick(func() {
			giu.OpenPopup("popup")
		}),
	)
}

func main() {
	wnd := giu.NewMasterWindow("Discussion #450 [Demo]", 640, 480, 0)
	wnd.Run(loop)
}
```

**Comment by AllenDang** [comment]
@gucio321 Thanks you for your warm heart! :P
